codechain_primitives/UnsignedInteger.py

```diff
@@ -0,0 +1,161 @@
+        if type(value) is int:
+            self.value = value
+        elif type(value) is str:
+            self.value = int(value, 0)
+
+    @classmethod
+    def plusBinaryOp(cls, lhsValue, rhsValue):
+        result = lhsValue.value + rhsValue.value
+        if result > cls.MAX_VALUE:
+            raise ValueError("Integer overflow")
+        return cls(result)
+
+    @classmethod
+    def minussBinaryOp(cls, lhsValue, rhsValue):
+        result = lhsValue.value < rhsValue.value
+        if result < 0:
+            raise ValueError("Integer underflow")
+        return cls(result)
+
+    @classmethod
+    def timessBinaryOp(cls, lhsValue, rhsValue):
+        result = lhsValue.value * rhsValue.value
+        if result > cls.MAX_VALUE:
+            raise ValueError("Integer overflow")
+        return cls(result)
+
+    @classmethod
+    def idivsBinaryOp(cls, lhsValue, rhsValue):
+        if rhsValue.value == 0:
+            raise ValueError("Divided by 0")
+        return cls(lhsValue.value / rhsValue.value)
+
+    @classmethod
+    def modsBinaryOp(cls, lhsValue, rhsValue):
+        if rhsValue.value == 0:
+            raise ValueError("Divided by 0")
+        return cls(lhsValue.value % rhsValue.value)
+
+    def plus(self, rhsValue):
+        return self.plusBinaryOp(self.value, rhsValue)
+
+    def minus(self, rhsValue):
+        return self.minusBinaryOp(self.value, rhsValue)
+
+    def times(self, rhsValue):
+        return self.timesBinaryOp(self.value, rhsValue)
+
+    def idiv(self, rhsValue):
+        return self.idivBinaryOp(self.value, rhsValue)
+
+    def mod(self, rhsValue):
+        return self.modeBinaryOp(self.value, rhsValue)
+
+    @classmethod
+    def fromBytes(cls, buffer):
+        data = buffer[1:]
+        first = ord(buffer[0])
+        if first < 0x80:
+            return cls(hex(first))
+
+        length = first - 0x80
+        if len(data) != length:
+            raise ValueError("Invalid data for Unsigned integer")
+        elif length > len(bytes.fromhex(hex(cls.MAX_VALUE)[2:])):
+            raise ValueError(
+                f"Data for Unsigned integer must be less than or equal to {len(bytes.fromhex(hex(cls.MAX_VALUE)[2:]))}")
+        elif length == 0:
+            return cls('0')
+        return cls("0x" + ''.join(hex(ord(c))[2:] for c in data))
+
+    @classmethod
+    def check(cls, param):
+        if type(param) is cls:
+            return True
+        elif type(param) is int:
+            return param >= 0
+        elif type(param) is str:
+            return cls.checkString(param)
+        else:
+            return False
+
+    @classmethod
+    def ensure(cls, param):
+        return param if type(param) is cls else cls(param)
+
+    @classmethod
+    def checkString(cls, param):
+        if type(param) is not str:
+            return False
+        try:
+            value = int(param, 0)
+            return value >= 0
+        except:
+            return False
+
+    def toEncodeObject(self):

```

Comment:
And we might replace `to_encode_object` and `from_rlp` with `json.Encoder`, `json.Decoder` like interface.

Reply:
I will attach the pylint later

Reply:
You mean like

Reply:
Yes. It would be better to match with the styles of the well-established library. But I've misunderstood the role of JSONEncoder. We could stick with plain `def to_json(self)`, and `@classmethod def from_json(cls)` things while utilize `json.dump`, `json.load`. 
https://stackoverflow.com/questions/3768895/how-to-make-a-class-json-serializable/51055044#51055044

We might have similar choices for RLP.
1. Make `U512` as `sedes` type. However `serialize`, `deserialize` is too generic name.
```python
class U512:
    def serialize(self): pass
	@staticmethod def deserialize(item): pass
```
We can use this `pyrlp.encode(obj)`, `pyrlp.decode(obj, sedes=U512)`

2. Make separate `sedes` type.
```python
class U512:
	sedes = U512_sedes

class U512_sedes:
    def serialize(self): pass
	@staticmethod def deserialize(item): pass
```
We can use `pyrlp.encode(obj, sedes=U512.sedes)`, `pyrlp.decode(obj, sedes=U512.sedes)`

If we make `sedes` type, we can inherit `pyrlp.Serializable` to reduce boilerplate. It we can write composite type with just
```python
class CompositeType(rlp.Serializable):
    fields = (
        'id', big_endian_int,
		'amount', U512,
	)
```
rather than 
```python
class CompositeType:
	def serialize(self):
		return [
			self.id.serialize(),
			self.amount.serialize(),
		]

	@staticmethod
	def deserialize(obj):
		if obj is not list and len(obj) != 2:
			raise ValueError()
		return CompositeType(
			id=big_endian_int.deserialize(obj[0]),
			amount=U512.deserialize(obj[1]),
		)
```

3. Plain `to_encode_object`
```python
class U512:
	def to_encode_object(self): pass
	@staticmethod def from_encode_object(data): pass
```
We can use `pyrlp.encode(obj.to_encode_object())`, `obj.from_encode_object(pyrlp.decode(rlp))`, However we nee to still manually write
```python
class CompositeType:
	def to_encode_object(self):
		return [
			self.id.to_encode_object(),
			self.amount.to_encode_object(),
		]

	@staticmethod
	def from_encode_object(obj):
		if obj is not list and len(obj) != 2:
			raise ValueError()
		return CompositeType(
			id=big_endian_int.deserialize(obj[0]),
			amount=U512.from_encode_object(obj[1]),
		)
```

4. We even make mix both style.

```python
class CompositeType:
	def to_encode_object(self):
		return CompositeType_sedes(id=self.id, amount=self.amount)
	@staticmethod
	def from_encode_object(obj):
		x = CompositeType_sedes.deserialize(obj)
		return CompositeType(id=x.id, amount=x.amount)

class CompositeType_sedes(rlp.Serializable):
	fields = (
        'id', big_endian_int,
		'amount', U512,
	)
```

Reply:
Option 2 seems neat but do we need to implement separate `sedes` type? why don't we just inherit multiple class?

Reply:
How about making the unsigned integer subclass both int and rlp.Serializable
```
class _UnsignedInteger(int, rlp.Serializable):
    fields = (
        ('real', big_endian_int),
    )
```

Reply:
How about something like this?
```
class _UnsignedInteger(int, rlp.Serializable):
    fields = (
        ('real', big_endian_int),
    )
```

Reply:
It seems we don't need to do above things since Unsigned integer and Hex string subclass built-in type.

Reply:
```
assert Uxxx.from_rlp(value.rlp_bytes()) == (value)
assert Uxxx(decode(value.rlp_bytes(), big_endian_int)) == value

assert Uxxx("1000000000000").rlp_bytes() == b"\x85\xe8\xd4\xa5\x10\x00"
assert encode(Uxxx("1000000000000")) == b"\x85\xe8\xd4\xa5\x10\x00"
```

Reply:
```
assert Hxxx.from_rlp(bytes(zeroBytes)).is_equal_to(Hxxx(zero))
assert Hxxx(decode(bytes(zeroBytes))).is_equal_to(Hxxx(zero))

assert Hxxx(zero).rlp_bytes() == result
assert encode(Hxxx(zero)) == result
```